src/nix/daemon.cc

```diff
@@ -294,45 +347,73 @@ static void daemonLoop()
     }
 }
 
+/**
+ * Forward a standard IO connection to the given remote store.
+ *
+ * We just act as a middleman blindly ferry output between the standard
+ * input/output and the remote store connection, not processing anything.
+ *
+ * Loops until standard input disconnects, or an error is encountered.
+ */
+static void forwardStdioConnection(RemoteStore & store) {
+    auto conn = store.openConnectionWrapper();
+    int from = conn->from.fd;
+    int to = conn->to.fd;
+
+    auto nfds = std::max(from, STDIN_FILENO) + 1;
+    while (true) {
+        fd_set fds;
+        FD_ZERO(&fds);
+        FD_SET(from, &fds);
+        FD_SET(STDIN_FILENO, &fds);
+        if (select(nfds, &fds, nullptr, nullptr, nullptr) == -1)
+            throw SysError("waiting for data from client or server");
+        if (FD_ISSET(from, &fds)) {
+            auto res = splice(from, nullptr, STDOUT_FILENO, nullptr, SSIZE_MAX, SPLICE_F_MOVE);
+            if (res == -1)
+                throw SysError("splicing data from daemon socket to stdout");
+            else if (res == 0)
+                throw EndOfFile("unexpected EOF from daemon socket");
+        }
+        if (FD_ISSET(STDIN_FILENO, &fds)) {
+            auto res = splice(STDIN_FILENO, nullptr, to, nullptr, SSIZE_MAX, SPLICE_F_MOVE);
+            if (res == -1)
+                throw SysError("splicing data from stdin to daemon socket");
+            else if (res == 0)
+                return;
+        }
+    }
+}
+
+/**
+ * Process a client connecting to us via standard input/output
+ *
+ * Unlike `forwardStdioConnection()` we do process commands ourselves in
+ * this case, not delegating to another daemon.
+ */
+static void processStdioConnection(ref<Store> store)
+{
+    FdSource from(STDIN_FILENO);
+    FdSink to(STDOUT_FILENO);
+    /* Auth hook is empty because in this mode we blindly trust the
+       standard streams. Limiting access to those is explicitly
+       not `nix-daemon`'s responsibility. */
+    processConnection(store, from, to, Trusted, NotRecursive);
```

Comment:
Instead of documenting a limitation, I think we could make it work.

```suggestion
    processConnection(store, from, to, store->isTrustedClient(), NotRecursive);
```

The point of this mode of operation is that we forward everything, so that should probably include informing the client about their trustedness correctly.

Reply:
I actually went with the first one for 3 reasons:

1. I hope this can be a pure refactor and not change any behavior.
2. I don't think that change is enough to warrant removing the note: in the case where the underlying store is a `LocalStore`, for example, it will blindly return `Trusted`. So standard streams + non-`RemoteStore` still means an invalidated client.
3. It is good to kindly ferry along the whether the "next" store trusts the client, but we could do that in *both* the stdio and non-stdio cases.

Reply:
Also I remembered that the way it works today is that the trusting of operations in `processConnection` is based solely on this parameter, but the trusting status returned to the client already takes `store->isTrustedClient()` into account.

That means in a chained daemon situation that even if an intermediate daemon trusts the the client (and doesn't block any request) the result of `store->isTrustedClient()` is still the intersection of the trust of each link in the chain.

I actually think that might be good semantics:

1. Don't block requests on behalf of another store further down the chain, let the first untrusting store do the blocking.
2. Do accurately report to the eventual client whether privileged requests will actually go all the way through.